File: app/src/main/java/com/worldventures/dreamtrips/wallet/service/command/record/SecureMultipleRecordsCommand.java
package com.worldventures.dreamtrips.wallet.service.command.record;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.dreamtrips.core.janet.dagger.InjectableAction;
import com.worldventures.dreamtrips.core.utils.tracksystem.AnalyticsInteractor;
import com.worldventures.dreamtrips.wallet.analytics.tokenization.ActionType;
import com.worldventures.dreamtrips.wallet.analytics.tokenization.TokenizationAnalyticsLocationCommand;
import com.worldventures.dreamtrips.wallet.analytics.tokenization.TokenizationCardAction;
import com.worldventures.dreamtrips.wallet.domain.entity.record.Record;
import com.worldventures.dreamtrips.wallet.domain.storage.disk.RecordsStorage;
import com.worldventures.dreamtrips.wallet.service.nxt.BaseMultipleRecordsCommand;
import com.worldventures.dreamtrips.wallet.service.nxt.DetokenizeMultipleRecordsCommand;
import com.worldventures.dreamtrips.wallet.service.nxt.NxtInteractor;
import com.worldventures.dreamtrips.wallet.service.nxt.TokenizeMultipleRecordsCommand;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.techery.janet.ActionState;
import io.techery.janet.Command;
import io.techery.janet.command.annotations.CommandAction;
import io.techery.janet.helper.ActionStateToActionTransformer;
import rx.Observable;
import rx.functions.Action1;

@CommandAction
public class SecureMultipleRecordsCommand extends Command<List<Record>> implements InjectableAction {

   @Inject NxtInteractor nxtInteractor;
   @Inject AnalyticsInteractor analyticsInteractor;
   @Inject RecordsStorage recordsStorage;

   private final List<Record> records;
   private final boolean secureForLocalStorage;
   private final boolean skipTokenizationErrors;

   @Nullable
   private final ActionType actionType;

   private SecureMultipleRecordsCommand(List<Record> records, boolean secureForLocalStorage,
         boolean skipTokenizationErrors, @Nullable ActionType actionType) {

      this.records = records;
      this.secureForLocalStorage = secureForLocalStorage;
      this.skipTokenizationErrors = skipTokenizationErrors;
      this.actionType = actionType;
   }

   @Override
   protected void run(CommandCallback<List<Record>> callback) throws Throwable {
      boolean offlineModeEnabled = recordsStorage.readOfflineModeState();

      if (offlineModeEnabled) {
         callback.onSuccess(records);
      } else {
         if (secureForLocalStorage) {
            tokenizeRecords(records).subscribe(callback::onSuccess, callback::onFail);
         } else {
            detokenizeRecords(records).subscribe(callback::onSuccess, callback::onFail);
         }
      }
   }

   private Observable<List<Record>> tokenizeRecords(List<? extends Record> records) {
      return nxtInteractor.tokenizeMultipleRecordsPipe()
            .createObservable(new TokenizeMultipleRecordsCommand(records, skipTokenizationErrors))
            .doOnNext(processResultForAnalytics())
            .compose(new ActionStateToActionTransformer<>())
            .map(Command::getResult);
   }

   private Observable<List<Record>> detokenizeRecords(List<? extends Record> records) {
      return nxtInteractor.detokenizeMultipleRecordsPipe()
            .createObservable(new DetokenizeMultipleRecordsCommand(records, skipTokenizationErrors))
            .doOnNext(processResultForAnalytics())
            .compose(new ActionStateToActionTransformer<>())
            .map(Command::getResult);
   }

   @NonNull
   private Action1<ActionState<? extends BaseMultipleRecordsCommand>> processResultForAnalytics() {
      return actionState -> {
         if (actionState.status == ActionState.Status.SUCCESS) {
            sendTokenizationAnalytics(actionState.action.getResult(), true);
         }
         if (actionState.status == ActionState.Status.FAIL || actionState.status == ActionState.Status.SUCCESS) {
            sendTokenizationAnalytics(actionState.action.getRecordsProcessedWithErrors(), false);
         }
      };
   }

   private void sendTokenizationAnalytics(List<Record> records, boolean success) {
      if (actionType == null || records.isEmpty()) return;

      Queryable.from(records).forEachR(recordWithError -> analyticsInteractor.walletAnalyticsCommandPipe()
            .send(new TokenizationAnalyticsLocationCommand(
                  TokenizationCardAction.from(recordWithError, success, actionType, secureForLocalStorage)
            )));
   }

   public static class Builder {

      private final List<Record> records = new ArrayList<>();
      private final boolean secureForLocalStorage;

      private boolean skipTokenizationErrors;
      @Nullable
      private ActionType actionType;

      private Builder(List<? extends Record> records, boolean secureForLocalStorage) {
         this.records.addAll(records);
         this.secureForLocalStorage = secureForLocalStorage;
      }

      public static Builder prepareRecordForLocalStorage(List<? extends Record> records) {
         return new Builder(records, true);
      }

      public static Builder prepareRecordForSmartCard(List<? extends Record> records) {
         return new Builder(records, false);
      }

      public Builder withAnalyticsActionType(ActionType actionType) {
         this.actionType = actionType;
         return this;
      }

      public Builder skipTokenizationErrors(boolean skipTokenizationErrors) {
         this.skipTokenizationErrors = skipTokenizationErrors;
         return this;
      }

      public SecureMultipleRecordsCommand create() {
         return new SecureMultipleRecordsCommand(records, secureForLocalStorage, skipTokenizationErrors, actionType);
      }

   }

}

File: wallet/src/main/java/com/worldventures/wallet/service/command/record/SecureMultipleRecordsCommand.kt
package com.worldventures.wallet.service.command.record

import com.worldventures.janet.injection.InjectableAction
import com.worldventures.wallet.analytics.tokenization.ActionType
import com.worldventures.wallet.analytics.tokenization.TokenizationAnalyticsLocationCommand
import com.worldventures.wallet.analytics.tokenization.TokenizationCardAction
import com.worldventures.wallet.domain.entity.record.Record
import com.worldventures.wallet.domain.storage.disk.RecordsStorage
import com.worldventures.wallet.service.WalletAnalyticsInteractor
import com.worldventures.wallet.service.nxt.BaseMultipleRecordsCommand
import com.worldventures.wallet.service.nxt.DetokenizeMultipleRecordsCommand
import com.worldventures.wallet.service.nxt.NxtInteractor
import com.worldventures.wallet.service.nxt.TokenizeMultipleRecordsCommand
import com.worldventures.wallet.util.WalletFeatureHelper
import io.techery.janet.ActionState
import io.techery.janet.Command
import io.techery.janet.command.annotations.CommandAction
import io.techery.janet.helper.ActionStateToActionTransformer
import rx.Observable
import java.util.ArrayList
import javax.inject.Inject

@CommandAction
class SecureMultipleRecordsCommand private constructor(private val records: List<Record>, private val secureForLocalStorage: Boolean,
                                                       private val skipTokenizationErrors: Boolean, private val actionType: ActionType?) : Command<List<Record>>(), InjectableAction {

   @Inject lateinit var nxtInteractor: NxtInteractor
   @Inject lateinit var analyticsInteractor: WalletAnalyticsInteractor
   @Inject lateinit var recordsStorage: RecordsStorage
   @Inject lateinit var featureHelper: WalletFeatureHelper

   @Throws(Throwable::class)
   override fun run(callback: Command.CommandCallback<List<Record>>) {
      val offlineModeEnabled = featureHelper.offlineModeState(recordsStorage.readOfflineModeState())

      if (offlineModeEnabled) {
         callback.onSuccess(records)
      } else {
         if (secureForLocalStorage) {
            tokenizeRecords(records).subscribe({ callback.onSuccess(it) }) { callback.onFail(it) }
         } else {
            detokenizeRecords(records).subscribe({ callback.onSuccess(it) }) { callback.onFail(it) }
         }
      }
   }

   private fun tokenizeRecords(records: List<Record>): Observable<List<Record>> {
      return nxtInteractor.tokenizeMultipleRecordsPipe()
            .createObservable(TokenizeMultipleRecordsCommand(records, skipTokenizationErrors))
            .doOnNext(processResultForAnalytics())
            .compose(ActionStateToActionTransformer())
            .map { it.result }
   }

   private fun detokenizeRecords(records: List<Record>): Observable<List<Record>> {
      return nxtInteractor.detokenizeMultipleRecordsPipe()
            .createObservable(DetokenizeMultipleRecordsCommand(records, skipTokenizationErrors))
            .doOnNext(processResultForAnalytics())
            .compose(ActionStateToActionTransformer())
            .map { it.result }
   }

   private fun processResultForAnalytics(): (ActionState<out BaseMultipleRecordsCommand>) -> Unit {
      return { actionState ->
         if (actionState.status == ActionState.Status.SUCCESS) {
            sendTokenizationAnalytics(actionState.action.result, true)
         }
         if (actionState.status == ActionState.Status.FAIL || actionState.status == ActionState.Status.SUCCESS) {
            sendTokenizationAnalytics(actionState.action.getRecordsProcessedWithErrors(), false)
         }
      }
   }

   private fun sendTokenizationAnalytics(records: List<Record>, success: Boolean) {
      if (actionType == null || records.isEmpty()) {
         return
      }

      records.forEach { recordWithError ->
         analyticsInteractor.walletAnalyticsPipe()
               .send(TokenizationAnalyticsLocationCommand(
                     TokenizationCardAction.from(recordWithError, success, actionType, secureForLocalStorage)
               ))
      }
   }

   class Builder private constructor(records: List<Record>, private val secureForLocalStorage: Boolean) {

      private val records = ArrayList<Record>()

      private var skipTokenizationErrors: Boolean = false
      private var actionType: ActionType? = null

      init {
         this.records.addAll(records)
      }

      fun withAnalyticsActionType(actionType: ActionType): Builder {
         this.actionType = actionType
         return this
      }

      fun skipTokenizationErrors(skipTokenizationErrors: Boolean): Builder {
         this.skipTokenizationErrors = skipTokenizationErrors
         return this
      }

      fun create(): SecureMultipleRecordsCommand {
         return SecureMultipleRecordsCommand(records, secureForLocalStorage, skipTokenizationErrors, actionType)
      }

      companion object {

         fun prepareRecordForLocalStorage(records: List<Record>): Builder {
            return Builder(records, true)
         }

         fun prepareRecordForSmartCard(records: List<Record>): Builder {
            return Builder(records, false)
         }
      }

   }

}
